File: sms-api-java/userservice/src/main/java/com/example/mydemoproject/userservice/UserServiceApplication.java
package com.example.mydemoproject.userservice;

import lombok.extern.slf4j.Slf4j;
import org.modelmapper.ModelMapper;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.cloud.client.discovery.EnableDiscoveryClient;
import org.springframework.context.annotation.Bean;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

@SpringBootApplication
@EnableDiscoveryClient
public class UserServiceApplication {
    public static void main(String[] args) {
        SpringApplication.run(UserServiceApplication.class, args);

        // functional programming
        List<Integer> numbers = List.of(1, 2, 3);
        List<Integer> numbersDoubled = map(numbers, x -> x * 2);

        System.out.println(numbersDoubled);

        //smart cast
        String stringText = "Hello world!";
        String nullableText = null;

        printStringLength(stringText);
        printStringLength(nullableText);

        // ternary operator
        int x = 2;

        String result = (x + 2 == 4) ? "yes" : "no";
        System.out.println("X equals 2 - " + result);

        // if - switch
        int dayOfWeek = 3;
        String dayName;

        switch (dayOfWeek) {
            case 1:
                dayName = "Monday";
                break;
            case 2:
                dayName = "Tuesday";
                break;
            case 3:
                dayName = "Wednesday";
                break;
            case 4:
                dayName = "Thursday";
                break;
            case 5:
                dayName = "Friday";
                break;
            default:
                dayName = "Weekend";
                break;
        }

        System.out.println("It's " + dayName);
    }

    public static <T, R> List<R> map(List<T> list, Function<T, R> function){
        List<R> result = new ArrayList<>();
        for(T item : list){
            result.add(function.apply(item));
        }
        return result;
    }

    public static void printStringLength(String str) {
        if (str != null) {
            // No smart cast in Java, using instanceof and type casting
            System.out.println("String length: " + str.length());
        } else {
            System.out.println("String is null");
        }
    }

    @Bean
    public ModelMapper modelMapper(){
        return new ModelMapper();
    }


}

File: sms-api-kotlin/userservice/src/main/kotlin/com/example/userservice/UserserviceApplication.kt
package com.example.userservice

import lombok.extern.slf4j.Slf4j
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.runApplication
import org.springframework.cloud.client.discovery.EnableDiscoveryClient

@SpringBootApplication
@EnableDiscoveryClient
class UserserviceApplication

fun main(args: Array<String>) {
	runApplication<UserserviceApplication>(*args)
	// functional programming
	val numbers = listOf(1, 2, 3)
	val numbersDoubled = numbers.map { it * 2 }
	println(numbersDoubled)

	//smart cast
	val stringText: String? = "Hello world!"
	val nullableText: String? = null

	printStringLength(stringText)
	printStringLength(nullableText)

	// ternary operator doesn't exist
	val x = 2

	val result = if (x + 2 == 4) "yes" else "no"
	println("X equals 2 - $result")

	//if - when
	val dayOfWeek = 3

	val dayName = when (dayOfWeek) {
		1 -> "Monday"
		2 -> "Tuesday"
		3 -> "Wednesday"
		4 -> "Thursday"
		5 -> "Friday"
		else -> "Weekend"
	}

	println("It's $dayName")
}

fun printStringLength(str: String?) {
	if (str != null) {
		// Smart cast: Compiler knows that str is not null inside this block
		println("String length: ${str.length}")
	} else {
		println("String is null")
	}
}